File: plot_boxplots_all_maps.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
plot_boxplots_all_maps.py — 论文主图：5 张地图 × 10 算法箱线图（matplotlib）。

与 MATLAB 版同源数据：records_merged.csv；LLM-EEFO → EEFOLLM。

纵轴指标（--metric）：
  bestfit_feasible（默认）— 仅 CollisionFree=1 的 run 上的 BestFit（与常见论文里「十几～几十」的路径代价量级一致，
    不含碰撞硬惩罚 1e4 量级）。
  bestfit — 全部 run 的 BestFit（含碰撞项，可达数千～数万，用于展示失败/碰撞时的完整目标）。
  path_length — 路径长度（PathLength 列，网格单位，通常几十～百余）。

说明：若某列 IQR=0，箱高为 0 属正常。对数纵轴按数据跨度自动选择；零方差列加粗中位线。

依赖：numpy, matplotlib
  pip install numpy matplotlib

用法（仓库根目录）：
  python plot_boxplots_all_maps.py
  python plot_boxplots_all_maps.py --metric bestfit
  python plot_boxplots_all_maps.py --csv path/to/records.csv
"""

from __future__ import annotations

import argparse
import csv
import sys
from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np

ALGOS10 = [
    "AOA",
    "EO",
    "GTO",
    "HHO",
    "HO",
    "MPA",
    "SBOA",
    "SMA",
    "SSA",
    "EEFOLLM",
]
MAPS = ["Map1", "Map2", "Map3", "Map4", "Map5"]
EXPECTED_RUNS = 20
LOG_RATIO_THRESH = 50.0


def repo_root() -> Path:
    return Path(__file__).resolve().parent


def load_records(csv_path: Path):
    rows = []
    with csv_path.open(newline="", encoding="utf-8") as f:
        for row in csv.DictReader(f):
            a = row.get("Algorithm", "").strip()
            if a == "LLM-EEFO":
                a = "EEFOLLM"
            row["Algorithm"] = a
            rows.append(row)
    return rows


def _collision_free(row: dict) -> bool:
    v = row.get("CollisionFree", "")
    if v is None or v == "":
        return False
    try:
        return float(v) >= 0.5
    except (TypeError, ValueError):
        return str(v).strip().lower() in ("1", "true", "yes")


def build_matrix(
    rows: list[dict],
    map_name: str,
    algos: list[str],
    expected_runs: int,
    metric: str = "bestfit_feasible",
) -> np.ndarray:
    """Shape (expected_runs, n_algo), rows = run index 1..expected_runs."""
    na = len(algos)
    M = np.full((expected_runs, na), np.nan, dtype=float)
    for ai, algo in enumerate(algos):
        sub = [r for r in rows if r.get("Map") == map_name and r.get("Algorithm") == algo]
        for r in sub:
            try:
                run = int(float(r.get("Run", "nan")))
            except (TypeError, ValueError):
                continue
            if not (1 <= run <= expected_runs):
                continue
            try:
                if metric == "bestfit":
                    M[run - 1, ai] = float(r["BestFit"])
                elif metric == "bestfit_feasible":
                    if _collision_free(r):
                        M[run - 1, ai] = float(r["BestFit"])
                elif metric == "path_length":
                    M[run - 1, ai] = float(r["PathLength"])
                else:
                    M[run - 1, ai] = float(r["BestFit"])
            except (KeyError, TypeError, ValueError):
                pass
    return M


def metric_ylabel_title_suffix(metric: str) -> tuple[str, str]:
    if metric == "bestfit":
        return (
            "Fitness value (full objective, incl. collision penalty)",
            " — full objective",
        )
    if metric == "path_length":
        return ("Path length (grid units)", " — path length")
    return (
        "Best fitness (collision-free runs only)",
        " — feasible runs",
    )


def pick_y_scale(M: np.ndarray, mode: str) -> bool:
    if mode == "log":
        return True
    if mode == "linear":
        return False
    v = M[np.isfinite(M) & (M > 0)]
    if v.size < 2:
        return False
    r = float(np.max(v) / np.min(v))
    return np.isfinite(r) and r >= LOG_RATIO_THRESH


def plot_one_map(
    M: np.ndarray,
    map_name: str,
    algos: list[str],
    use_log: bool,
    out_paths: list,
    metric: str,
) -> None:
    n = len(algos)
    fig_w = max(12.0, 0.95 * n + 2.0)
    fig, ax = plt.subplots(figsize=(fig_w, 5.2), dpi=120)
    fig.patch.set_facecolor("white")

    # 列数据列表（matplotlib boxplot 要 list of 1d）
    cols = [M[:, j].astype(float) for j in range(n)]
    cols_plot = []
    for c in cols:
        x = c[np.isfinite(c) & (c > 0)]
        cols_plot.append(x if x.size else np.array([np.nan]))

    positions = np.arange(1, n + 1)
    vpos = M[np.isfinite(M) & (M > 0)]

    if vpos.size == 0:
        ax.text(
            0.5,
            0.5,
            "No usable Y values for this metric on this map.\n"
            "For --metric bestfit_feasible: no collision-free runs.\n"
            "Try:  python plot_boxplots_all_maps.py --metric bestfit",
            transform=ax.transAxes,
            ha="center",
            va="center",
            fontsize=10,
            color="0.35",
        )
    else:
        ax.boxplot(
            cols_plot,
            positions=positions,
            widths=0.55,
            patch_artist=True,
            showfliers=True,
            flierprops=dict(marker="+", markerfacecolor="red", markersize=7, linestyle="none"),
            medianprops=dict(color="0.05", linewidth=2.2),
            whiskerprops=dict(color="0.35", linewidth=1.0, linestyle="--"),
            capprops=dict(color="0.35", linewidth=1.0),
            boxprops=dict(facecolor="#dbe0ee", edgecolor="#4a5560", linewidth=1.0),
        )

        # 零方差列：加粗中位线区段（避免“看不见箱体”时仍有一条细线）
        for j, c in enumerate(cols):
            arr = c[np.isfinite(c) & (c > 0)]
            if arr.size < 2:
                continue
            u = np.unique(arr)
            spread = float(np.max(arr) - np.min(arr))
            if u.size == 1 or spread < 1e-12 * max(1.0, float(np.median(arr))):
                m = float(np.median(arr))
                ax.plot(
                    [positions[j] - 0.38, positions[j] + 0.38],
                    [m, m],
                    color="#1a3a5c",
                    linewidth=5.0,
                    solid_capstyle="butt",
                    zorder=4,
                    alpha=0.85,
                )

        if use_log:
            ax.set_yscale("log")
            lo = float(np.nanmin(vpos))
            hi = float(np.nanmax(vpos))
            if np.isfinite(lo) and np.isfinite(hi) and lo > 0:
                if hi <= lo * 1.001:
                    hi = lo * 1.15
                ax.set_ylim(lo * 0.88, hi * 1.12)
        else:
            lo = float(np.nanmin(M[np.isfinite(M)]))
            hi = float(np.nanmax(M[np.isfinite(M)]))
            if np.isfinite(lo) and np.isfinite(hi):
                pad = (hi - lo) * 0.06 + 1e-9
                if pad < 1e-12:
                    pad = max(abs(lo), 1.0) * 0.05
                ax.set_ylim(lo - pad, hi + pad)

        # 仅「可行 run」指标下：无碰撞自由路径的算法在数据上就是全 NaN，不是漏画
        if metric == "bestfit_feasible":
            lo_a, hi_a = ax.get_ylim()
            if ax.get_yscale() == "log":
                y_note = float(np.sqrt(lo_a * hi_a))
            else:
                y_note = (lo_a + hi_a) / 2
            for j in range(n):
                c = cols[j]
                good = np.isfinite(c) & (c > 0)
                if np.any(good):
                    continue
                ax.text(
                    positions[j],
                    y_note,
                    "No feasible\npath (0/20)",
                    ha="center",
                    va="center",
                    fontsize=7.5,
                    color="0.4",
                    zorder=10,
                )

    ax.set_xticks(positions)
    ax.set_xticklabels(algos, rotation=40, ha="right")
    ax.grid(True, which="major", linestyle="-", alpha=0.35)
    ax.grid(True, which="minor", linestyle=":", alpha=0.2)
    ax.set_axisbelow(True)

    ylab, title_suffix = metric_ylabel_title_suffix(metric)
    if use_log:
        ylab += " (log-scaled Y; lower is better)"
    else:
        ylab += " (lower is better)"
    ax.set_ylabel(ylab, fontsize=11)
    ax.set_title(
        f"{map_name} — distribution over 20 independent runs{title_suffix}",
        fontsize=12,
        fontweight="normal",
    )

    plt.tight_layout()
    for p in out_paths:
        p = Path(p)
        p.parent.mkdir(parents=True, exist_ok=True)
        if p.suffix.lower() == ".png":
            fig.savefig(p, dpi=300, bbox_inches="tight", facecolor="white")
        elif p.suffix.lower() == ".svg":
            try:
                fig.savefig(p, format="svg", bbox_inches="tight", facecolor="white")
            except Exception as e:
                print(f"Warning: SVG save failed {p}: {e}", file=sys.stderr)
    plt.close(fig)


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--csv", type=str, default="", help="records CSV path")
    ap.add_argument(
        "--y-scale",
        choices=("auto", "log", "linear"),
        default="auto",
        help="Y axis scale policy",
    )
    ap.add_argument(
        "--metric",
        choices=("bestfit_feasible", "bestfit", "path_length"),
        default="bestfit_feasible",
        help="Y value: feasible BestFit (default, ~10^1), full BestFit (large if collisions), or PathLength",
    )
    args = ap.parse_args()

    root = repo_root()
    if args.csv:
        csv_path = Path(args.csv)
    else:
        csv_path = root / "results" / "global_experiments_merged" / "tables" / "records_merged.csv"
    if not csv_path.is_file():
        print(f"Error: CSV not found: {csv_path}", file=sys.stderr)
        return 1

    rows = load_records(csv_path)
    fig_dir = root / "figures" / "paper_main"
    res_dir = root / "results" / "boxplot_maps"
    fig_dir.mkdir(parents=True, exist_ok=True)
    res_dir.mkdir(parents=True, exist_ok=True)

    metric_tag = {
        "bestfit_feasible": "fitness",
        "bestfit": "fitness_full",
        "path_length": "pathlength",
    }[args.metric]

    for map_name in MAPS:
        M = build_matrix(rows, map_name, ALGOS10, EXPECTED_RUNS, args.metric)
        use_log = pick_y_scale(M, args.y_scale)
        base = f"boxplot_{metric_tag}_{map_name}"
        base2 = f"boxplot_{metric_tag}_{map_name}_all10"
        outs = [
            fig_dir / f"{base}.png",
            fig_dir / f"{base}.svg",
            res_dir / f"{base2}.png",
            res_dir / f"{base2}.svg",
        ]
        plot_one_map(M, map_name, ALGOS10, use_log, outs, args.metric)
        print(f"Saved: {base}.png/.svg -> figures/paper_main + results/boxplot_maps")

    print("Done.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: plot_boxplots_all_maps.m
function plot_boxplots_all_maps(opts)
%PLOT_BOXPLOTS_ALL_MAPS 五张 benchmark 地图各一张箱线图：10 算法 × 20 次独立实验最终 BestFit。
%
%   数据来源与 generate_main_tables 一致：records_merged.csv 或 batch1~3 合并；LLM-EEFO → EEFOLLM。
%   每列原始 20 次 BestFit（不足则为 NaN），不使用均值代替。
%
%   当不同算法最优值跨度很大（如 MPA 与其它方法差多个数量级）时，默认使用对数纵轴以便比较；
%   可用 opts.y_scale='linear' 强制线性坐标（此时个别方法会拉高纵轴，其它箱体被压扁属正常现象）。
%   红色「+」仅为箱线图规则下的离群点；无离群点则不会显示。
%   跨度大时不再依赖 YScale=log（易被 boxplot/导出打回线性刻度），改为对数据取 log10 后在线性 Y 上作图，
%   纵轴为 log10(fitness)，箱体与分位数均在 log10 空间计算。
%
%   输出（每张地图各一份 PNG + SVG）：
%     figures/paper_main/boxplot_fitness_Map1..Map5
%     results/boxplot_maps/boxplot_Map1..Map5_all10
%
%   用法（工程根目录）：
%     plot_boxplots_all_maps
%     plot_boxplots_all_maps(struct('records_csv', 'path/to/records.csv', 'y_scale', 'log'))
%
%   若 MATLAB 导出/坐标异常，推荐改用 Python（matplotlib，对数 Y + 散点叠加）：
%     python plot_boxplots_all_maps.py

if nargin < 1, opts = struct(); end
if ~isfield(opts, 'expected_runs'), opts.expected_runs = 20; end
if ~isfield(opts, 'out_dir'), opts.out_dir = ''; end
if ~isfield(opts, 'records_csv'), opts.records_csv = ''; end
if ~isfield(opts, 'font_name'), opts.font_name = 'Times New Roman'; end
if ~isfield(opts, 'font_size'), opts.font_size = 9; end
if ~isfield(opts, 'y_scale'), opts.y_scale = 'auto'; end %#ok<*STRQUOT>
if ~isfield(opts, 'log_ratio_thresh'), opts.log_ratio_thresh = 50; end
if ~isfield(opts, 'tick_angle'), opts.tick_angle = 40; end
if ~isfield(opts, 'fig_width_px'), opts.fig_width_px = 1280; end
if ~isfield(opts, 'fig_height_px'), opts.fig_height_px = 520; end

root_dir = fileparts(mfilename('fullpath'));
addpath(root_dir);
addpath(fullfile(root_dir, 'config'));
addpath(fullfile(root_dir, 'utils'));
cfg = default_config();
init_paths(cfg.paths.root);

if isempty(opts.out_dir)
    opts.out_dir = fullfile(cfg.paths.figures, 'paper_main');
end
ensure_dir(opts.out_dir);
res_box_dir = fullfile(cfg.paths.results, 'boxplot_maps');
ensure_dir(res_box_dir);

[algos10, ~] = paper_benchmark_algorithms10();
maps = cfg.maps.names(:);
if isstring(maps), maps = cellstr(maps); end

T = load_records_for_boxplot(opts, cfg);
T = normalize_algo_boxplot(T);
T.Algorithm = cellstr(string(T.Algorithm));
T.Map = cellstr(string(T.Map));

exp_maps = maps(:);
exp_algos = algos10(:);
T = T(ismember(string(T.Map), string(exp_maps)) & ismember(string(T.Algorithm), string(exp_algos)), :);

ylab_base = 'Fitness value';
nG = numel(exp_algos);
% 轮廓式箱线图 + 深色线（勿用 BoxStyle=filled：exportgraphics 出 SVG 时常把浅灰填充/描边
% 画成极淡的 line，白底上几乎看不见；PNG 上亦可能类似）
lineRgb = [0.22 0.32 0.45];
boxColors = repmat(lineRgb, nG, 1);

lbls = cellfun(@char, exp_algos, 'UniformOutput', false);

for k = 1:5
    mn = exp_maps{k};
    M = build_bestfit_matrix(T, mn, exp_algos, opts.expected_runs);

    useLog = pick_y_scale(M, opts);
    ylab = ylab_base;
    if useLog
        ylab = [ylab_base, ' (log10 scale; lower is better)'];
    end

    f = figure('Color', 'w', 'Units', 'pixels', ...
        'Position', [80 80 opts.fig_width_px opts.fig_height_px]);
    ax = axes('Parent', f);
    Mplot = M;
    if useLog
        Mplot(Mplot <= 0) = NaN;
        Mplot = log10(Mplot);
    end

    % 使用 'Parent' 兼容旧版；轮廓线宽、略加宽箱体
    h = boxplot(Mplot, 'Parent', ax, 'Labels', lbls, ...
        'Symbol', 'r+', 'OutlierSize', 5, 'Whisker', 1.5, ...
        'Colors', boxColors, 'Widths', 0.55);
    if ~isempty(h)
        try
            set(h, 'LineWidth', 1.05);
        catch
        end
    end
    % 统一加深所有箱线元素（部分版本 Colors 未覆盖全部句柄）
    darken_boxplot_lines(ax, lineRgb);

    grid(ax, 'on');
    box(ax, 'on');
    ylabel(ax, ylab, 'FontName', opts.font_name, 'FontSize', opts.font_size + 1);
    title(ax, sprintf('%s — distribution of final fitness (20 independent runs)', mn), ...
        'FontName', opts.font_name, 'FontSize', opts.font_size + 2, 'FontWeight', 'normal');
    set(ax, 'FontName', opts.font_name, 'FontSize', opts.font_size);
    set(ax, 'TickLabelInterpreter', 'none');
    xtickangle(ax, opts.tick_angle);

    % 为斜置标签留出下边距
    set(ax, 'Position', [0.10 0.20 0.86 0.68]);

    if useLog
        vm = Mplot(isfinite(Mplot));
        if ~isempty(vm)
            ylim(ax, [min(vm(:)) - 0.18, max(vm(:)) + 0.18]);
        end
        set(ax, 'YTickMode', 'auto');
    end
    style_boxplot_medians(ax);

    drawnow;
    baseFig = fullfile(opts.out_dir, sprintf('boxplot_fitness_%s', mn));
    baseRes = fullfile(res_box_dir, sprintf('boxplot_%s_all10', mn));
    export_fig_vector_raster_box(f, baseFig);
    export_fig_vector_raster_box(f, baseRes);
    close(f);

    fprintf(1, 'Saved: %s / %s (PNG+SVG)\n', baseFig, baseRes);
end
end

function useLog = pick_y_scale(M, opts)
useLog = false;
ys = char(string(opts.y_scale));
switch lower(strtrim(ys))
    case 'log'
        useLog = true;
    case 'linear'
        useLog = false;
    case 'auto'
        % isfinite(M) 与 M 同尺寸；勿与 M(:)>0 混用（列向量与矩阵尺寸不兼容）
        v = M(isfinite(M) & M > 0);
        if numel(v) < 2
            return;
        end
        r = max(v) / min(v);
        useLog = isfinite(r) && (r >= opts.log_ratio_thresh);
    otherwise
        useLog = false;
end
end

function darken_boxplot_lines(ax, rgb)
% 将箱线图中的 line 统一为可见深色（避免浅灰线在白纸/矢量导出中消失）
L = findall(ax, 'Type', 'line');
for i = 1:numel(L)
    t = get(L(i), 'Tag');
    if iscell(t), t = t{1}; end
    t = char(string(t));
    % 保留离群点为红色；其余（含无 Tag 的箱线）统一加深；中位数随后再改为黑色加粗
    if strcmpi(strtrim(t), 'Outliers')
        continue;
    end
    try
        set(L(i), 'Color', rgb);
    catch
    end
end
end

function style_boxplot_medians(ax)
hm = findall(ax, 'Tag', 'Median');
if isempty(hm), hm = findall(ax, 'Tag', 'median'); end
if ~isempty(hm)
    set(hm, 'Color', [0.05 0.05 0.05], 'LineWidth', 1.85);
end
end

function M = build_bestfit_matrix(T, map_name, algos, expected_runs)
na = numel(algos);
M = nan(expected_runs, na);
for ai = 1:na
    ix = strcmp(T.Map, map_name) & strcmp(T.Algorithm, algos{ai});
    sub = T(ix, :);
    for r = 1:expected_runs
        j = find(double(sub.Run) == r, 1);
        if ~isempty(j)
            M(r, ai) = sub.BestFit(j);
        end
    end
end
end

function T = load_records_for_boxplot(opts, cfg)
T = table();
if ~isempty(opts.records_csv) && isfile(opts.records_csv)
    T = readtable(opts.records_csv, 'TextType', 'string');
    return;
end
p = fullfile(cfg.paths.results, 'global_experiments_merged', 'tables', 'records_merged.csv');
if isfile(p)
    T = readtable(p, 'TextType', 'string');
    return;
end
try
    T = merge_global_experiment_batches();
catch
    T = merge_batches_mem_box(cfg);
end
end

function T = merge_batches_mem_box(cfg)
parts = {};
for b = 1:3
    f = fullfile(cfg.paths.results, sprintf('global_experiments_batch%d', b), 'tables', 'records.csv');
    if isfile(f)
        parts{end+1} = readtable(f, 'TextType', 'string'); %#ok<AGROW>
    end
end
if isempty(parts), T = table(); return; end
T = vertcat(parts{:});
end

function T = normalize_algo_boxplot(T)
a = string(T.Algorithm);
a(a == "LLM-EEFO") = "EEFOLLM";
T.Algorithm = cellstr(a);
end

function export_fig_vector_raster_box(f, base_path)
png = [base_path, '.png'];
svg = [base_path, '.svg'];
if exist('exportgraphics', 'file')
    exportgraphics(f, png, 'Resolution', 300);
    try
        exportgraphics(f, svg, 'ContentType', 'vector');
    catch
        try
            saveas(f, svg);
        catch
            warning('plot_boxplots:NoSvg', 'SVG export failed; PNG only saved.');
        end
    end
else
    print(f, png, '-dpng', '-r300');
    try
        print(f, svg, '-dsvg');
    catch
    end
end
end
